client/pull.go

```diff
@@ -152,20 +183,37 @@ func (c *Client) downloadFilePart(ctx context.Context, dst *os.File, url string,
 }
 
 // downloadWorker is a worker func for processing jobs in stripes channel.
-func (c *Client) downloadWorker(ctx context.Context, dst *os.File, url string, parts <-chan partSpec, pb ProgressBar) func() error {
+func (c *Client) downloadWorker(ctx context.Context, dst *os.File, endpoint, authToken string, parts <-chan partSpec, pb ProgressBar) func() error {
 	return func() error {
 		for ps := range parts {
-			if err := c.downloadFilePart(ctx, dst, url, &ps, pb); err != nil {
+			if err := c.downloadFilePart(ctx, dst, endpoint, authToken, &ps, pb); err != nil {
 				return err
 			}
 		}
 		return nil
 	}
 }
 
-func (c *Client) getContentLength(ctx context.Context, url string) (int64, error) {
+// parseContentRangeHeader returns size returned in Content-Range response HTTP header
+func parseContentRangeHeader(value string) (int64, error) {
+	if value == "" {
+		return -1, nil
+	}
+
+	vals := strings.Split(value, "/")
+	if len(vals) < 2 {
+		return 0, errUnknownContentLength
+	}
+	if vals[1] == "*" {
+		// Server reports size is unknown
+		return 0, fmt.Errorf("indeterminant size")
+	}
+	return strconv.ParseInt(vals[1], 0, 64)
+}
+
+func (c *Client) getContentLength(ctx context.Context, endpoint, authToken string) (int64, error) {
```

Comment:
General question (applies to old code same as new).... is there a reason this uses a `GET` and looks at `Content-Range`, rather than doing a `HEAD` and looking at `Content-Length`?

Reply:
At least in the case of an externally redirected URL from cloud library 1.x to MinIO, (curiously...) a `HEAD` request results in the `Content-Length` header value set to `0` (as well as an unexpected 403 Forbidden HTTP status). The same pre-signed S3 endpoint called with `GET` reports the length correctly.

`HEAD`

```sh
$ curl -I https://XXXX/singularity-enterprise-library/bson.007230/6387986149ab6b512d032700\?X-Amz-Algorithm\=AWS4-HMAC-SHA256\&X-Amz-Credential\=changeme%2F20230110%2Fus-east-1%2Fs3%2Faws4_request\&X-Amz-Date\=20230110T133833Z\&X-Amz-Expires\=3600\&X-Amz-SignedHeaders\=host\&X-Amz-Signature\=5b1d0b858cb29b1630a7e05b85b670f1188af98f2dc00ed8633c2de3b40f4dfd
HTTP/2 403
date: Tue, 10 Jan 2023 13:41:06 GMT
content-length: 0
accept-ranges: bytes
content-security-policy: block-all-mixed-content
strict-transport-security: max-age=31536000; includeSubDomains
vary: Origin
vary: Accept-Encoding
x-amz-request-id: 1738F64BA840C58D
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
```

`GET`



```sh
$ curl -o /dev/null https://XXXXX/singularity-enterprise-library/bson.007230/6387986149ab6b512d032700\?X-Amz-Algorithm\=AWS4-HMAC-SHA256\&X-Amz-Credential\=changeme%2F20230110%2Fus-east-1%2Fs3%2Faws4_request\&X-Amz-Date\=20230110T133833Z\&X-Amz-Expires\=3600\&X-Amz-SignedHeaders\=host\&X-Amz-Signature\=5b1d0b858cb29b1630a7e05b85b670f1188af98f2dc00ed8633c2de3b40f4dfd

<snipped for brevity>

< HTTP/2 200
< date: Tue, 10 Jan 2023 13:47:16 GMT
< content-type: application/octet-stream
< content-length: 337940480
< accept-ranges: bytes
< content-security-policy: block-all-mixed-content
< etag: "e13d67e86176f6054b0de819ffea6680"
< last-modified: Wed, 30 Nov 2022 17:53:00 GMT
< no-gzip-compression: true
< strict-transport-security: max-age=31536000; includeSubDomains
< vary: Origin
< vary: Accept-Encoding
< x-amz-request-id: 1738F6A1BE8075A1
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< x-amz-meta-client-md5sum: e13d67e86176f6054b0de819ffea6680
```

The GET request is simply used to reliably determine if the remote server supports range requests.

Reply:
Fair enough. That's pretty surprising behaviour to me. I would have expected `accept-ranges` and `content-length` to be set appropriately for a `HEAD` request, such that GET 'ing a range isn't needed.

I guess these S3 servers continue to surprise. Sorry for the noise!